### PR TITLE
Provide DockerContainer implementation

### LINKED ISSUE #2089: Implement a Container abstraction
This aims to implement a container abstraction. That is, for now we need `DockerContainer` but the abstraction should facilitate plugin in whatever container driver we may want to use. This will facilitate work on other container providers as well as make Docker-less testing and simulation possible.

### PR DESCRIPTION
**Pull request by sven-lange-last**
This implementation is an abstraction layer for using Docker containers as OpenWhisk action containers - in preparation of a new container pool implementation (#2086). The `DockerContainer` implementation provides a high-level abstraction for the operations required by OpenWhisk:

- Create, destroy, halt, resume a container
- Initialize and run an action inside a container
- Obtain container logs

Closes #2089

### CODE REVIEW

File: core/invoker/src/main/scala/whisk/core/containerpool/Container.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+     * Initializes code in the container.
+     */
+    def initialize(initializer: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[Interval]
+
+    /**
+     * Runs code in the container.
+     */
+    def run(parameters: JsObject, environment: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[(Interval, ActivationResponse)]
+
+    def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Future[Vector[String]]
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/Container.scala]: Scaladoc please.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import whisk.common.LoggingMarkers
+import whisk.core.entity.ByteSize
+import java.nio.charset.StandardCharsets
+import whisk.http.Messages
+import whisk.core.entity.ActivationResponse
+import whisk.core.containerpool.BlackboxStartupError
+import whisk.core.containerpool.WhiskContainerStartupError
+import whisk.core.containerpool.InitializationError
+import whisk.core.containerpool.Container
+import whisk.core.invoker.ActionLogDriver
```
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: care to reorg imports?

Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: Done - also taken care of imports in other files...

File: core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala [hunk 3]
```diff
@@ -50,7 +50,7 @@ import whisk.core.entity.size.SizeLong
  * @param timeoutMsec the timeout in msecs to wait for a response
  * @param maxResponse the maximum size in bytes the connection will accept
  */
-protected[container] class HttpUtils(
+class HttpUtils(
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala]: why making this generally visible?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        val actionName = environment.fields.get("action_name").map(_.convertTo[String]).getOrElse("")
+        val start = transid.started(this, LoggingMarkers.INVOKER_ACTIVATION_RUN, s"sending arguments to $actionName at $id $ip")
+
+        val parameterWrapper = JsObject("value" -> parameters)
+        val body = JsObject(parameterWrapper.fields ++ environment.fields)
+        callContainer("/run", body, timeout, retry = false).andThen { // never fails
+            case Success(r: RunResult) =>
+                transid.finished(this, start.copy(start = r.interval.start), s"running result: ${r.toBriefString}", endTime = r.interval.end)
+            case Failure(t) =>
+                transid.failed(this, start, s"initializiation failed with $t")
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: initializiation -> should be run

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+        val environmentArgs = (environment + ("SERVICE_IGNORE" -> true.toString)).map {
+            case (key, value) => Seq("-e", s"$key=$value")
+        }.flatten
+
+        val args = Seq(
+            "--cap-drop", "NET_RAW",
+            "--cap-drop", "NET_ADMIN",
+            "--ulimit", "nofile=1024:1024",
+            "--pids-limit", "1024", // OW PR 2119
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: 2119 already merged; remove comment.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+            "--cpu-shares", cpuShares.toString,
+            "--memory", s"${memory.toMB}m",
+            "--memory-swap", s"${memory.toMB}m",
+            "--network", network) ++
+            environmentArgs ++
+            name.map(n => Seq("--name", n)).getOrElse(Seq.empty)
+
+        val pulled = if (userProvidedImage) {
+            docker.pull(image).recoverWith {
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: we should put a limit on the docker pull operation (per previous discussion with @perryibm this was not done in the past so as penalize for connectivity to dockerhub - but the absence of any time limit is also generally bad).

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+     * @param userProvidedImage whether the image is provided by the user
+     *     or is an OpenWhisk provided image
+     * @param memory memorylimit of the container
+     * @param cpuShares sharefactor for the container
+     * @param environment environment variables to set on the container
+     * @param network network to launch the container in
+     * @param name optional name for the container
+     * @return a container
+     */
+    def create(transid: TransactionId,
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: add return type to the definition.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+     * @param transid transaction creating the container
+     * @param image image to create the container from
+     * @param userProvidedImage whether the image is provided by the user
+     *     or is an OpenWhisk provided image
+     * @param memory memorylimit of the container
+     * @param cpuShares sharefactor for the container
+     * @param environment environment variables to set on the container
+     * @param network network to launch the container in
+     * @param name optional name for the container
+     * @return a container
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: returns a future which either completes with a container or one of two specific failures.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+ *
+ * This class contains OpenWhisk specific behavior and as such does not necessarily
+ * use docker commands to achieve the effects needed.
+ *
+ * @constructor
+ * @param id the id of the container
+ * @param ip the ip of the container
+ */
+class DockerContainer(id: ContainerId, ip: ContainerIp)(
+    implicit docker: DockerApiWithFileAccess, runc: RuncApi, ec: ExecutionContext, log: Logging) extends Container with ActionLogDriver {
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: i suggest renaming `log` to `logger` or something that more obviously connotes this is the system logger.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+                Future.failed(InitializationError(ActivationResponse.applicationError(Messages.timedoutActivation(timeout, true)), result.interval))
+            } else {
+                Future.failed(InitializationError(ActivationResponse.processInitResponseContent(result.response, log), result.interval))
+            }
+        }
+    }
+
+    def run(parameters: JsObject, environment: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[(Interval, ActivationResponse)] = {
+        val actionName = environment.fields.get("action_name").map(_.convertTo[String]).getOrElse("")
+        val start = transid.started(this, LoggingMarkers.INVOKER_ACTIVATION_RUN, s"sending arguments to $actionName at $id $ip")
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: is the action name logging important here?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        val parameterWrapper = JsObject("value" -> parameters)
+        val body = JsObject(parameterWrapper.fields ++ environment.fields)
+        callContainer("/run", body, timeout, retry = false).andThen { // never fails
+            case Success(r: RunResult) =>
+                transid.finished(this, start.copy(start = r.interval.start), s"running result: ${r.toBriefString}", endTime = r.interval.end)
+            case Failure(t) =>
+                transid.failed(this, start, s"initializiation failed with $t")
+        }.map { result =>
+            val response = if (result.interval.duration >= timeout) {
+                ActivationResponse.applicationError(Messages.timedoutActivation(timeout, false))
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: this is a corner case here where duration = timeout and result.response is valid.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+            val conn = new HttpUtils(s"${ip.asString}:8080", timeout, 1.MB)
+            httpConnection = Some(conn)
+            conn
+        }
+        Future {
+            http.post(path, body, retry)
+        }.map { response =>
+            val finished = Instant.now()
+            RunResult(Interval(started, finished), response)
+        }
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: how about a recover for defense sake and log an error; or assert a violation.

File: tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+        docker.inspects should have size 1
+        docker.rms should have size 0
+
+        val (testImage, args) = docker.runs.head
+        testImage shouldBe "image"
+
+        // Assert fixed values are passed as well
+        args should contain allOf ("--cap-drop", "NET_RAW", "NET_ADMIN")
+        args should contain inOrder ("--ulimit", "nofile=1024:1024")
+        args should contain inOrder ("--pids-limit", "1024") // OW PR 2119
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala]: remove 2119 comment

Comment (rabbah) [inline tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala]: could consolidate all these into a `Seq(x, y).foreach args should contain inOrder (x, y)`

File: tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala [hunk 14]
```diff
@@ -0,0 +1,640 @@
...
+
+        docker.pulls should have size 0
+        docker.runs should have size 1
+        docker.inspects should have size 1
+        docker.rms should have size 1
+    }
+
+    it should "disambiguate errors if user images are provided" in {
+        implicit val docker = new TestDockerClient {
+            override def inspectIPAddress(id: ContainerId, network: String)(implicit transid: TransactionId): Future[ContainerIp] = {
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala]: this test seems wrong in that unlike the next test that follows, failure on inspect is a system error not a black box error, is it not?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 10 above)
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: It is for some performance related tooling. This keeps this log as-is.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 11 above)
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: You're saying it should be `duration > timeout`? I think this is semantically preserving the current code, but worth double-checking.

File: tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala [hunk 15]
```diff
@@ -77,8 +77,8 @@ class ActionLogDriverTests
             raw"""|{"time":"","stream":"stdout","log":"a"}
                   |{"time":"","stream":"stdout","log":"b"}
                   |{"time":"","stream":"stdout","log":"c"}
-                  |{"time":"","stream":"stdout","log":"$LOG_ACTIVATION_SENTINEL\n"}
-                  |{"time":"","stream":"stderr","log":"$LOG_ACTIVATION_SENTINEL\n"}""".stripMargin('|')
+                  |{"time":"","stream":"stdout","log":"$ActionLogDriver.LOG_ACTIVATION_SENTINEL\n"}
+                  |{"time":"","stream":"stderr","log":"$ActionLogDriver.LOG_ACTIVATION_SENTINEL\n"}""".stripMargin('|')
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala]: How do those still pass? Doesn't this need to be `${ActionLogDriver.LOG_ACTIVATION_SENTINEL}`

File: core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala (hunk 3 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala]: `whisk.core.containerpool.docker.DockerContainer` uses `HttpUtils` so it cannot be `protected[container]`. Changed to `protected[core]`.

File: tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala (hunk 15 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala]: Good finding - changed to `${ActionLogDriver.LOG_ACTIVATION_SENTINEL}`. Also removed the explicit line endings `\n` because further processing removes them anyway.

The problem with `${ActionLogDriver.LOG_ACTIVATION_SENTINEL}` wasn't detected by unit tests because those don't cover sentinel processing.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 12 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: I'm not sure what you have in mind. At the moment, we have two consumers of `callContainer()`, i.e. `initialize()` and `run()`. Those two consumers deal with a failed `Future`.

If we add a `recover` here and return a successful `Future[RunResult]` in any case, the caller's error handling isn't used any more.

So should we keep the error handling as is - or move it down to `callContainer()`?

Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: The clients don't have recovers, so they don't deal with a failing future.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: the contract from the http client is that it should never fail the future - so my comment was to defensively code against an incompatible change there (or, to make sure we have a test to assert the call always returns a successful future).

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 11 above)
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: likely preserving behavior - i trust you'll check. but if so the behavior is wrong for the border case anyway. either fix or open issue to fix later.

File: tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala (hunk 15 above)
Comment (rabbah) [inline tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala]: so is the test wrong? i'm not sure i understood the last comment.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 11 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: I checked - it's in `getActivationResponse()` in `Invoker.scala`:

```
        if (interval.duration >= timeout) {
            ActivationResponse.applicationError(Messages.timedoutActivation(timeout, failedInit))
        } else ...
```

I opt for keeping the current semantics even if we have the impression that the corner case `duration == timeout` may need special handling. I think that we should keep behaviour changes with the new container pool at a minimum so that we can compare test results. We can open an issue and follow up once the new container pool is in place.

File: tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala (hunk 15 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/invoker/ActionLogDriverTests.scala]: The `ActionLogDriver` unit test isn't wrong - but doesn't fully cover the area of sentinel processing. Our goal was to keep the `ActionLogDriver` and the existing unit tests as is. So we aligned our new log processing code around `ActionLogDriver` - but externalized the `LOG_ACTIVATION_SENTINEL` such that we can use it in our new unit tests for `DockerClientWithFileAccess`.

In summary, no urgent need for changes in `ActionLogDriver` and corresponding unit tests.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 12 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: Besides the access scope change, we left `HttpUtils` as is. It still returns a `Left()` if unsuccessful or a `Right()` if successful.

Both callers of the code under discussion here - namely, `initialize()` and `run()` - have the following error handling code:

```
        callContainer("/init", body, timeout, retry = true).andThen { // never fails
            case Success(r: RunResult) =>
                transid.finished(this, start.copy(start = r.interval.start), s"initialization result: ${r.toBriefString}", endTime = r.interval.end)
            case Failure(t) =>
                transid.failed(this, start, s"initializiation failed with $t")
        } ...
```

So we are going to log a failed `Future` returned by `callContainer()` and return the failed `Future` to the caller of `initialize()` and `run()`. Looking at https://github.com/openwhisk/openwhisk/pull/2021/files, I do NOT think that the caller of `initialize()` and `run()` fully addresses a failed `Future`.

@rabbah do you suggest to `recover` here at line 251 and transform the `Failure(t)` into a `RunResult(..., Left(t))`? This transformation would bypass the code snippet from above with `transid.failed()`. I guess we need a decision how hard the penalty for a failed `Future` is: a log statement, a failed transaction marker, a failed init / run, or even an assertion stopping the invoker process. What do you have in mind?

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 6 above)
Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: I agree - `docker pull` takes potentially very long and may block a single Docker container for a long time. It's reasonable to let the operation fail if pulling an image takes too long.

- What would be a reasonable timeout duration for `docker pull`? Should we re-use the same timeout we apply to initialize / run? Or do we need a general `docker pull` timeout?
- At the moment, I'm building a prototype with a timeout mechanism based on `akka.pattern.after`. Unfortunately, said `after()` needs a scheduler - most likely implying that we need the proper actor system at this point. So instead of passing the `ExecutionContext`, we need the whole `ActorSystem` and use `ActorSystem.scheduler` and `ActorSystem.dispatcher`.
- Said approach with `akka.pattern.after` used in `DockerContainer.create()` will let the creation fail with a timeout if it takes too long - but won't actually abort the running `docker pull` command process. Our current command execution approach does not support aborting a running process. We may think about providing a process executor that observes timeouts and abort execution upon timeout.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 12 above)
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: As we return a Future[Either] we might aswell leave it as-is. The documentation I think states the contract already. If we want to leave HttpUtils as-is I'd leave that wrinkle in.

Alternatively we could flatten the Future[Either] where for the Left case we fail the Future. Note though that we leave type-information in that case.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: @sven-lange-last are you able to restore the scope of the `HttpUtils` to `container`? If so, we could do nothing beyond the comment. I am a little concerned that earlier `protected[container]` became `public` where we might end up making an incompatible change that we don't catch.

File: core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala (hunk 6 above)
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: we have a timeout feature we use in the controller that may be reusable here see `ExecutionContextFactory` for the future and promise extensions that offer `after`.

Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: We should not use the init and run timeout - unless we also treat this as a billable operation. Perhaps we should.

Comment (sven-lange-last) [inline core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala]: We created a variant with timeout in `ProcessRunner` that will be provided in a separate PR.

File: tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala (hunk 14 above)
Comment (sven-lange-last) [inline tests/src/test/scala/whisk/core/containerpool/docker/test/DockerContainerTests.scala]: You are right. We inspect the Docker config to obtain the IPv4 address iff `docker run` has been successful. This inspection is expected to work properly even if the container has exited meanwhile.
For this reason, an inspection failure should not be a `BlackboxStartupError`which is mapped to an application error.

**Comment by sven-lange-last** [comment]
PG3 - 594 🔵
